### DISCUSSION TITLE
What's the best rust entrypoint for transforming?

### DISCUSSION THREAD

**Discussion by brocoppler** [swc-project/swc#1992] [Q&A] [answered]
I'm trying to use JNI to transform some JS from a Java project. I was using the `process_js_file` function in `src/lib.rs` and passing in my own JS file and `.swcrc` file like so:
``` 
let fm = cm
        .load_file(Path::new("/tmp/foo.js"))
        .expect("failed to load file");
let transformed = c
        .process_js_file(
            fm,
            &Options {
                config_file: Some(ConfigFile::Str("/tmp/.swcrc".to_string())),
                ..Default::default()
            },
        )
        .expect("failed to process file");
```

And this seems to work... sort of. There are some unfortunate things I can't seem to get around like decorators are left out, the use strict directive is never included and modules just plain don't work. Is there a better API I should be calling? Thanks!

**Comment by kdy1** [comment, marked as answer]
It depends on your usecase. Deno uses`swc_ecma_transfoms` via `swc_ecmascript`, and next uses `swc` crate.

**Reply by brocoppler** [reply]
Thanks for the quick response!
Basic use case is that I want to take ES6+ code, transform it to ES5 & consume the code + source map. What makes it more complex is that I'm doing this all from Java so I need to interact with swc through JNI. I saw the example in here using `process_js_file` but it seems like that isn't a great entry point since it doesn't seem to do a full transform no matter what I do. 

Is `swc_ecmascript` a separate crate/project?

**Reply by brocoppler** [reply]
I see, it is a separate crate. What is the API to use for `swc_ecma_transforms`? I think that may be my best bet

**Reply by kdy1** [reply]
The documentation for visitors at https://rustdoc.swc.rs/swc_visit/  will help.
You can see list of visitor implementations in https://docs.rs/swc_ecma_transforms/0.61.0/swc_ecma_transforms/

**Reply by brocoppler** [reply]
Which function should I start at if I just want to pass in some JavaScript & have it return ES5 JS & source maps?

**Reply by kdy1** [reply]
There is no such function. swc does really lots of transforms, and single function does not fit.

**Reply by brocoppler** [reply]
Thanks, this was all helpful. I ended up using the `transform_sync` method from `wasm` since it was the simplest for my use case

**Comment by caoccao** [comment]
It might be too late for this question. You might want to check [swc4j](https://github.com/caoccao/swc4j) out.
